### ISSUE TITLE
Write a dynamism RFC

### ISSUE THREAD

**Issue by burmako** [openxla/stablehlo#8]
MHLO's design for supporting dynamic shapes is pretty battle-tested already, and we've inherited it in StableHLO by bootstrapping from MHLO. Let's further iterate on it to address some of the rough edges, e.g. one possibility would be to merge dynamic ops like `dynamic_pad` with regular ops like `pad`. We already had some discussions about dynamism with [the TCP effort](https://discourse.llvm.org/t/rfc-proposal-for-a-high-level-ml-dialect-in-mlir/64249), and I'm looking forward to continuing that.

**Comment by burmako** [comment]
I haven't gotten a chance to make progress with this ticket yet. This week, I was focusing on unblocking adoption of StableHLO by integrating it into MLIR-HLO, so that existing MHLO producers / consumers have an easy way to experiment with it. Next week, I'm planning to work on compatibility guarantees (#4), and then I'm hoping to get to this ticket next.

**Comment by burmako** [comment]
I'll need to spend some more time on integration work (the recently opened ticket on integration with TF, JAX, PyTorch/XLA and Torch-MLIR), but I'm aiming to start working on dynamism in early Q4. I have a pretty fleshed out draft, but haven't yet had the time to finish and publish it.

**Comment by burmako** [comment]
"I'm aiming to start working on dynamism in early Q4. I have a pretty fleshed out draft, but haven't yet had the time to finish and publish it". In early Q4, I got involved in speccing statically-shaped ops, and this was a very interesting experience. The more I worked on the specs, the more I realized that the dynamism RFC should be written differently that what I imagined before.

Previously, I planned to approach the dynamism RFC similarly to #194 - describe the baseline, summarize design goals, write down which fields will be added / changed in which ops, provide some examples, and that's it. However, when writing and reviewing specs, I observed that this approach can leave a lot of corner cases out of sight / out of mind. Given that we are not blocked on the dynamism RFC (the design of dynamism that we inherited from MHLO has many known issues but it works), I'm hoping to approach this differently.

In December, I expect that two things will happen: 1) we will finish speccing statically-shaped StableHLO ops, 2) we will formalize syntax and evaluation of StableHLO programs (#484). Once that's done, I'm planning to write the dynamism RFC as the delta to this formalism (where the usual design document will be accompanied with changes to spec.md). I think this won't be too hard and will result in a higher-quality design, because we'll be forced to explore more details.

Moreover, having a formalism in place will allow us to prove things about StableHLO programs. E.g. specifying the conditions under which type inference can specialize dynamically-shaped programs to static shapes. Or resolving open questions like #229. I think this is worth investing the extra time.

**Comment by burmako** [comment]
In this comment, I will be keeping track of stakeholder feedback about the current state of dynamism in MHLO/StableHLO:
  1. The only user of unranked dynamism at the moment is TensorFlow, and it can handle it in a different layer, so we shouldn't need unranked tensors in StableHLO.
  2. Same for dynamic axes, we shouldn't need them in StableHLO (note that axes, i.e. dimension numbers, are different from dimension sizes - clearly, we need dynamic dimension sizes).
  3. Having different code paths for producing/consuming static and dynamic ops (FooOp vs DynamicFooOp) is a testing/maintenance/cognitive burden ([example](https://github.com/llvm/torch-mlir/pull/1331)).
  4. Shapes as `tensor<?xindex>` introduce brittleness into the whole flow, and they are not mandatory without unranked dynamism ([example 1](https://github.com/llvm/torch-mlir/pull/1306), [example 2](https://github.com/iree-org/iree/issues/10334)).
  5. At the moment, indexes in StableHLO are modelled inconsistently. For example, SortOp requires an `i64` dimension, but GetTupleElementOp requires an `i32` element number.
  6. There is broad alignment on shape mismatches being UB, but there are multiple schools of thought on how shape mismatches should be guarded against.
  7. Closure under automatic differentiation and vectorization is important.
  8. How should type inference work in the presence of dynamic types? For some of the open questions, see #229, #621 and #622.
  9. Maybe broadcast ops could support negative indices in `broadcast_dimensions`? This would improve expressivity in unranked cases. (Cf. with the bullets above where we discussed that unranked dynamism is something that we shouldn't need in StableHLO).
  10. Does DynamicBroadcastInDim really need the ability to dynamically decide whether to broadcast dimensions or not? Also see [[Public] TCP Design - Broadcast](https://docs.google.com/document/d/1x8kCUsYzFEIeWO38h32bHlmDrieeu9UJTxpKe9gUm_g/edit#).
  11. How does DynamicConvOp work? Why does it have both `d_padding` and `padding`? The name of the former is not ideal.
  12. Do we need both DynamicSliceOp and RealDynamicSliceOp?

**Comment by miladm** [comment]
Had offline chats with @smit-hinsu. 

Logically speaking, we may want to break up the discussion into three segments - see below. Happy to hear what other folks have in mind.

+ What's our bounded dynamism API proposal?
+ What's our unbounded dynamism API proposal?
+ What's our proposed path to unify the bounded and unbounded API?

CC @burmako

**Comment by burmako** [comment]
Milad, thank you for the introduction! I also wanted to provide some additional context on StableHLO's plans for dynamism. In a nutshell these plans involve:
  1) Support for bounded dynamism (already informally exists in StableHLO, needs to be formalized).
  2) Support for unbounded dynamism (already informally exists in StableHLO, needs to be formalized).
  3) Potential unification.

1\) For bounded dynamism, [the RFC linked by Milad](https://github.com/openxla/stablehlo/pull/194) talks about the design in StableHLO which involves roughly the same functionality as HLO: a) bounds for dimension sizes in tensor types, b) get_dimension_size and set_dimension_size ops that map directly on HLO ops. In StableHLO, we take compatibility very seriously, and we want to keep providing these APIs for frontends like PyTorch/XLA that have investment in HLO-style dynamism.

2\) For unbounded dynamism, we're going to have a separate RFC in the near future, and the work on this RFC is tracked in https://github.com/openxla/stablehlo/issues/8. In the unbounded dynamism RFC, we'll be aiming to: 
  + Unify already existing dynamic_foo and foo ops (e.g. dynamic_reshape and reshape).
  + Provide clear principles by which we decide what metadata can be dynamic and what metadata must be static (e.g. should ReverseOp::dimensions be static or dynamic? what about SliceOp::*indices?).
  + Decide on the API to represent shapes (shape as a separate entity? shape as a vararg of dimension sizes? something else?).

3\) In the long term, we would also like to unify bounded and unbounded APIs, and [the RFC linked by Milad](https://github.com/openxla/stablehlo/pull/194) talks about some initial ideas in this area (see the "(P4) Aspirational: Migration to unbounded dynamism" section). This is some very early thinking, and I'm not yet sure how this will work out - it would be preferable to have just one API, but maybe we'll end up with two.

**Comment by burmako** [comment]
Here's the long awaited RFC: #1668.

**Comment by mlevesquedion** [comment]
The RFC has been merged: https://github.com/openxla/stablehlo/pull/1881

See it here: https://github.com/openxla/stablehlo/blob/main/rfcs/20230704-dynamism-101.md
